In [1]:
import pandas as pd
import numpy as np
from sklearn.metrics import confusion_matrix

In [2]:
# reading in results.csv
df = pd.read_csv('mini_batch_results.csv')

# drop useless columns
drop_cols = ['HITId', 'HITTypeId', 'Title', 'Description', 'Keywords', 'Reward', 'RequesterAnnotation', 'Approve', 'Reject',
             'NumberOfSimilarHITs', 'AutoApprovalDelayInSeconds', 'AssignmentDurationInSeconds', 'LifetimeInSeconds']
drop_cols.extend(col for col in df.columns if 'Answer.opt' in col)
df = df.drop(drop_cols, axis=1)
df

CreationTime  MaxAssignments  \
0    Mon Mar 02 22:08:57 PST 2020             129   
1    Mon Mar 02 22:08:57 PST 2020             129   
2    Mon Mar 02 22:08:57 PST 2020             129   
3    Mon Mar 02 22:08:57 PST 2020             129   
4    Mon Mar 02 22:08:57 PST 2020             129   
5    Mon Mar 02 22:08:57 PST 2020             129   
6    Mon Mar 02 22:08:57 PST 2020             129   
7    Mon Mar 02 22:08:57 PST 2020             129   
8    Mon Mar 02 22:08:57 PST 2020             129   
9    Mon Mar 02 22:08:57 PST 2020             129   
10   Mon Mar 02 22:08:57 PST 2020             129   
11   Mon Mar 02 22:08:57 PST 2020             129   
12   Mon Mar 02 22:08:57 PST 2020             129   
13   Mon Mar 02 22:08:57 PST 2020             129   
14   Mon Mar 02 22:08:57 PST 2020             129   
15   Mon Mar 02 22:08:57 PST 2020             129   
16   Mon Mar 02 22:08:57 PST 2020             129   
17   Mon Mar 02 22:08:57 PST 2020             129   
18   Mon Mar 02 22:08:57 PST 2020             129   
19   Mon Mar 02 22:08:57 PST 2020             129   
20   Mon Mar 02 22:08:57 PST 2020             129   
21   Mon Mar 02 22:08:57 PST 2020             129   
22   Mon Mar 02 22:08:57 PST 2020             129   
23   Mon Mar 02 22:08:57 PST 2020             129   
24   Mon Mar 02 22:08:57 PST 2020             129   
25   Mon Mar 02 22:08:57 PST 2020             129   
26   Mon Mar 02 22:08:57 PST 2020             129   
27   Mon Mar 02 22:08:57 PST 2020             129   
28   Mon Mar 02 22:08:57 PST 2020             129   
29   Mon Mar 02 22:08:57 PST 2020             129   
..                            ...             ...   
99   Mon Mar 02 22:08:57 PST 2020             129   
100  Mon Mar 02 22:08:57 PST 2020             129   
101  Mon Mar 02 22:08:57 PST 2020             129   
102  Mon Mar 02 22:08:57 PST 2020             129   
103  Mon Mar 02 22:08:57 PST 2020             129   
104  Mon Mar 02 22:08:57 PST 2020             129   
105  Mon Mar 02 22:08:57 PST 2020             129   
106  Mon Mar 02 22:08:57 PST 2020             129   
107  Mon Mar 02 22:08:57 PST 2020             129   
108  Mon Mar 02 22:08:57 PST 2020             129   
109  Mon Mar 02 22:08:57 PST 2020             129   
110  Mon Mar 02 22:08:57 PST 2020             129   
111  Mon Mar 02 22:08:57 PST 2020             129   
112  Mon Mar 02 22:08:57 PST 2020             129   
113  Mon Mar 02 22:08:57 PST 2020             129   
114  Mon Mar 02 22:08:57 PST 2020             129   
115  Mon Mar 02 22:08:57 PST 2020             129   
116  Mon Mar 02 22:08:57 PST 2020             129   
117  Mon Mar 02 22:08:57 PST 2020             129   
118  Mon Mar 02 22:08:57 PST 2020             129   
119  Mon Mar 02 22:08:57 PST 2020             129   
120  Mon Mar 02 22:08:57 PST 2020             129   
121  Mon Mar 02 22:08:57 PST 2020             129   
122  Mon Mar 02 22:08:57 PST 2020             129   
123  Mon Mar 02 22:08:57 PST 2020             129   
124  Mon Mar 02 22:08:57 PST 2020             129   
125  Mon Mar 02 22:08:57 PST 2020             129   
126  Mon Mar 02 22:08:57 PST 2020             129   
127  Mon Mar 02 22:08:57 PST 2020             129   
128  Mon Mar 02 22:08:57 PST 2020             129   

                       Expiration                    AssignmentId  \
0    Mon Mar 16 23:08:57 PDT 2020  31LM9EDVOPTATOLPAPQU5RJVMSDJNO   
1    Mon Mar 16 23:08:57 PDT 2020  31N2WW6R9VRNKI6J9GFUHFTXO06F3U   
2    Mon Mar 16 23:08:57 PDT 2020  31QNSG6A5VUW6MXJ3ECEK4OR5SS87B   
3    Mon Mar 16 23:08:57 PDT 2020  326O153BMMZTWW8K7U10C6U929YEDK   
4    Mon Mar 16 23:08:57 PDT 2020  32EYX73OY4A4M2OBFVBPLDXG3WAURA   
5    Mon Mar 16 23:08:57 PDT 2020  32RIADZISW5H6JU9FEKCYJ4IRVG4SH   
6    Mon Mar 16 23:08:57 PDT 2020  32ZKVD547JOX71U34ETYYP2C8ZGB33   
7    Mon Mar 16 23:08:57 PDT 2020  336YQZE83ZFW47L49M70R2BQYS75M4   
8    Mon Mar 16 23:08:57 PDT 2020  33IZTU6J85284LIJZPMHVU1GKSXSX0   
9    Mon Mar 16 23:08:57 PDT 2020  33JKGHPFY

In [3]:
def slider_to_array(slider_vals):
    """ Convert slider_vals from output.csv to a Python array. Slider_vals 
    is expected to be a string of numbers separated by ','. Assumes the last
    char in slider_vals is ','."""
    try:
        vals = slider_vals[:-1].split(',')
        vals = [int(x) for x in vals]
        return(vals)
    except:
        return([])
    
def calc_slider_std(row, window_size=15, shift=5):
    """Calculate the smallest standard deviation out of all the windows
    in each slider. For each slider, splits the values into windows 
    of size window_size (where
    the start index is separated by 'shift' spaces) and calculates the 
    standard deviation for each window. Takes all standards deviations
    of both sliders and returns the min."""
    stds = []
    for slider_vals in [row['Answer.slider_values'], row['Answer.slider_values2']]:
        vals = slider_to_array(slider_vals)
        if len(vals) == 0:
            return([])
        i = 0
        while i < len(vals):
            end = min(len(vals), i+window_size)
            stds.append(np.std(vals[i:end]))
            i += shift
    return(min(stds))
    
def calc_slider_extreme(row, extreme):
    """ Calculates the min or max of all slider values. 
    If extreme == 'min', returns min value of all slider values from worker.
    If extreme == 'max', returns max value of all slider values from worker."""
    all_vals = []
    for slider_vals in [row['Answer.slider_values'], row['Answer.slider_values2']]:
        slider_vals = slider_to_array(slider_vals)
        all_vals.extend(slider_vals)
    if len(all_vals) == 0:
        return(-1)
    if extreme == 'min':
        return(min(all_vals))
    elif extreme == 'max':
        return(max(all_vals))
    else:
        raise ValueError("Expected 'min' or 'max' for parameter extreme, received %s." % extreme)
    
def predict(row, params):
    """Predicts whether a person's results should be Approved or Rejected based on the specified
    params."""
    if row['Answer.srcc_score'] == 'initial':
        return('Rejected')
    srcc_score = pd.to_numeric(row['Answer.srcc_score'])
    std = row[std_col]
    slider_min = row['slider_min']
    slider_max = row['slider_max']

    
    if (std < params['std_thresh']) or (srcc_score <= params['srcc_thresh']) or (slider_min < params['min_thresh']) or (slider_max > params['max_thresh']):
        return('Rejected')
    return('Approved')

These indices have low standard deviation: [1,38,18,29,37, 40, 41, 47, 62, 67, 71, 77, 80, 88, 90, 99, 108, 111, 119] 

These indices have somewhat low standard deviation: [38, 39, 75, 118] 

These indices have responses that are too low: [69, 78, 120] 


In [4]:
# set these params
params = {
    'std_window': 15, # size of window used for calculating standard deviation
    'std_shift': 5, # how much to shift each window by (ex: window=15, shift=5 has windows [0, 14], [5, 19], [10, 24], ...)
    'std_thresh': 0.75, # lowest allowed standard deviation for a window
    'srcc_thresh': -0.5, # lowest allowed SRCC value
    'min_thresh': 4, # minimum allowed answer for any slider
    'max_thresh': 49, # maximum allowed answer for any slider
}

std_col = 'std_%i_%i' % (params['std_window'], params['std_shift'])
df[std_col] = df.apply(calc_slider_std, axis=1, window_size=params['std_window'], shift=params['std_shift']) # calc min std for all windows
df['slider_min'] = df.apply(calc_slider_extreme, axis=1, extreme='min') # calc min slider val
df['slider_max'] = df.apply(calc_slider_extreme, axis=1, extreme='max') # calc max slider val

df['predict'] = df.apply(predict, axis=1, params=params) # make predictions for approve/reject

In [5]:
# predictions: column 'predict'
# actual: column 'AssignmentStatus'
df_predict = df[['AssignmentStatus', 'predict', 'Answer.srcc_score', std_col, 'slider_min', 'slider_max']]
df_predict[df_predict['AssignmentStatus'] != df_predict['predict']]

AssignmentStatus   predict     Answer.srcc_score  std_15_5  slider_min  \
1           Rejected  Approved  0.051298917604257706  0.979796          30   
10          Approved  Rejected                  -0.5   2.93636          10   
17          Approved  Rejected                   0.8   2.41661          14   
37          Rejected  Approved     0.872081599272381    1.0198          11   
46          Approved  Rejected    0.6668859288553503  0.748331           9   
51          Approved  Rejected    0.6668859288553503   2.46216          10   
59          Approved  Rejected   0.30779350562554625       0.8          30   
67          Rejected  Approved    0.8207826816681233       0.8          12   
69          Rejected  Approved    0.5642880936468347   2.38886           4   
83          Approved  Rejected   0.41039134083406165   3.34398          22   
89          Approved  Rejected                  -0.5   2.24499           7   
99          Rejected  Approved   0.20519567041703082   1.13529           4   
114         Approved  Rejected                  -0.5       0.8          21   
118         Rejected  Approved   0.15389675281277312    1.0328           6   

     slider_max  
1            49  
10           48  
17           50  
37           34  
46           29  
51           50  
59           50  
67           25  
69           25  
83           50  
89           44  
99           30  
114          46  
118          26

In [6]:
# confusion matrix (1: approved, 0: rejected)
# TP: true positive: predicted = 1, actual = 1
# TN: true negative: predicted = 0, actual = 0
# FP: false positive: predicted = 1, actual = 0
# FN: false negative: predicted = 0, actual = 1
# [[TP, FP], [FN, TN]]
actual = df['AssignmentStatus']
predicted = df['predict']
confusion_matrix(actual, predicted, labels=['Approved', 'Rejected'])

array([[90,  8],
       [ 6, 25]], dtype=int64)

## Printing slider values for Rejected and Approved responses

In [7]:
# rejected slider values
df[df['AssignmentStatus'] == 'Rejected'][['Answer.slider_values', 'Answer.slider_values2', 'Answer.srcc_score', std_col]].style.set_properties(subset=['Answer.slider_values', 'Answer.slider_values2'], **{'width': '300px'})

In [8]:
# approved slider values
df[df['AssignmentStatus'] == 'Approved'][['Answer.slider_values', 'Answer.slider_values2', 'Answer.srcc_score', std_col]].style.set_properties(subset=['Answer.slider_values', 'Answer.slider_values2'], **{'width': '300px'})